Fix for QuasiHydrostatic with stagger time step

jm-c opened the pull request:
## What changes does this PR introduce?
Fix Adams-Bashforth time-stepping of QuasiHydrostatic terms when using staggerTimeStep.

## What is the current behaviour? 
 With Stagger time-step, since (Quasi) Hydrostatic pressure gradient is kept outside AB time-stepping (as it 
 should for pure buoyancy contribution), the quasi-hydro contributions (Coriolis-prime and NH metric-terms) 
 are advanced in time with just a (unstable) forward Euler-time-stepping.
This could be left unnoticed when QH terms are small (vs typical ocean stratification) but could results
in fatal instability when QH terms are large with very weak/no stratification. 

## What is the new behaviour 
Fix Adams-Bahforth time-stepping of Quasi-Hydrostatic terms for staggerTImeStep by applying AB extrapolation
in time on QH tendency terms. 

## Does this PR introduce a breaking change? 
Need storage of previous time-step tendency (in common block and in pickup files for restart) ; to avoid this
if not used, put storage within CPP option:  ALLOW_QHYD_STAGGER_TS (default: undef).

## Other information:
Tested in homogeneous 40.km deep ocean on small planet set-up.
Note: some addition ordering terms have changed in S/R MOM_QUASIHYDROSTATIC resulting
in numerical differences (at machine truncation level) even when using synchronous time-stepping.

## Suggested addition to `tag-index`
o model/src + pkg/mom_common:
  Fix AB time-stepping of QuasiHydrostatic with staggerTimeStep.
  - involves new CPP option "ALLOW_QHYD_STAGGER_TS" for storage of previous
   time-step Quasi-Hydro tendency to enable AB extrapolation in time when using staggerTimeStep.
 - field added to pickup file for restart ; allows to restart (but not perfect restart) if field is missing
   and pickupStrictlyMatch=F.
 - switch to staggerTimeStep=T in experiment "global_ocean.90x40x15.dwnslp" that uses QuasiHydrostatic
   to test this code; update corresponding output (as well as primary test output due to machine truncation).

mjlosch commented on model/src/read_pickup.F:
```diff
@@ -21,14 +21,12 @@ SUBROUTINE READ_PICKUP(
 #include "PARAMS.h"
 #include "RESTART.h"
 #include "DYNVARS.h"
+#include "NH_VARS.h"
```
This appears to be only necessary, when ALLOW_QHYD_STAGGER_TS or ALLOW_NONHYDROSTATIC defined. Use #ifdef/#endif?

jm-c replied:
You are right. But since 1) NH_VARS.h is always available (model/inc always in the path of src files) and 
2)  NH_VARS.h is empty when none of of the 2 are defined, it does not hurt too much. I've just thought it would make read_pickup.F simpler to follow.

mjlosch replied:
OK

mjlosch commented on model/src/write_pickup.F:
```diff
@@ -23,14 +23,12 @@ SUBROUTINE WRITE_PICKUP(
 #include "PARAMS.h"
 #include "RESTART.h"
 #include "DYNVARS.h"
+#include "NH_VARS.h"
```
same as with read_pickup: use #ifdef/#endif?

jm-c replied:
Same comment as for read_pickup.F ; I could add these double ifdef but not sure it would help.

mjlosch replied:
OK

mjlosch reviewed (approved):
Not sure what I can review here. Everything looks good to me, two small comments/questions that do not require any action. On my MacBook I get this with gfortran:
```
Y Y Y Y>10<13 16 16 16 16 16 16 16 16 13 13 13 13 13 11 14  7  7 14 16  .  .  .  . pass  global_ocean.90x40x15
Y Y Y Y>11<16 16 16 16 16 16 16 16 16 16 16 16 13 13 12 16 16 16 16 16  .  .  .  . pass  global_ocean.90x40x15.dwnslp
Y Y Y Y>12<16 16 16 16 16 16 16 16 16 13 12 16 14 13 12 14 -- -- -- --  .  .  .  . pass  global_ocean.90x40x15.idemix
```
which is probably OK. I don't remember the details of the QH approximation, but everything sounds plausible and looks good.

jm-c commented:
@mjlosh regarding testreport numbers, with PGI compiler (& -devel) I was getting (without the changes from this PR):
Y Y Y Y>10<12 13 16 16 16 16 16 16 13 13 13 13 13 13 11 16 12 12 14 16  .  .  .  . pass  global_ocean.90x40x15
Y Y Y Y>10<13 16 16 16 16 16 16 16 13 13 13 13 13 13 11 14 16 16 16 16  .  .  .  . pass  global_ocean.90x40x15.dwnslp
Y Y Y Y>11<13 16 16 16 16 16 16 16 13 13 12 13 13 13 12 13 -- -- -- --  .  .  .  . pass  global_ocean.90x40x15.idemix
just to get a vague idea of what we could expect from different compiler.

jm-c commented:
@jrscott I added the new CPP option to the table in the docs, and also added ALLOW_SMAG_3D_DIFFUSIVITY which was missing (I am the one to blame for this). If you could take a quick look, would be great.
A random comment regarding this CPP_OPTIONS.h table: May be if we could find a way to separate these options into groups
(as done, with comments, in the file model/inc/CPP_OPTIONS.h) it could make it easier to digest. 
The current sections (from grep) are:
C-- Forcing code options:
C-- Options to discard parts of the main code:
C-- Vertical mixing code options:
C-- Time-stepping code options:
C-- Model formulation options:
C-- Algorithm options:
C-- Retired code options:
C-- Other option files:
Not that it should be in this PR but something we could think about in a future one.

jrscott commented:
@jm-c I'll take a look and consider options for the CPP_OPTIONS table.

jm-c commented:
@jrscott OK, but if you "consider options for the CPP_OPTIONS table" I think it would be better in a specific documentation PR
(and I should probably have opened an issue for that).

jrscott commented:
agree @jm-c, have made new issue on breaking up tables.

This PR: doc LGTM
Ready to merge?

jm-c commented:
@jrscott Thanks for the review and opening this new issue (it's better this way).
Will probably merge this PR tomorrow (since a significant PR has been merged today, better to wait 1 night to get full set of daily testreport output).
